mode/haskell-literate/haskell-literate.js

```diff
@@ -0,0 +1,42 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
+// Distributed under an MIT license: http://codemirror.net/LICENSE
+
+(function (mod) {
+  if (typeof exports == "object" && typeof module == "object") // CommonJS
+    mod(require("../../lib/codemirror"), require("../haskell/haskell"))
+  else if (typeof define == "function" && define.amd) // AMD
+    define(["../../lib/codemirror", "../haskell/haskell"], mod)
+  else // Plain browser env
+    mod(CodeMirror)
+})(function (CodeMirror) {
+  CodeMirror.defineMode("haskell-literate", function (config) {
+    var haskellMode = CodeMirror.getMode(config, "haskell")
+    return {
+      startState: function () {
+        return {
+          haskellCode: false,
+          haskellState: CodeMirror.startState(haskellMode)
+        }
+      },
+      copyState: function (state) {
+        return {
+          haskellState: CodeMirror.copyState(haskellMode, state.haskellState)
+        }
+      },
+      token: function (stream, state) {
+        var s = stream.next()
+        if (s == '>' || state.haskellCode) {
+          state.haskellCode = true
+          return haskellMode.token(stream, state.haskellState)
+        } else {
+          while (!stream.eol())
+            stream.next()
+          return "comment"
+        }
+      },
+      blankLine: function (state) {
+        state.haskellCode = false
+      }
+    }
+  })
```

Comment:
Modes that wrap other modes need to implement an [`innerMode`](http://codemirror.net/doc/manual.html#innerMode) method, to make things like completion and folding aware of the actual mode.